src/wasi.zig

```diff
@@ -651,9 +651,48 @@ pub fn integrationTest() void {
         return;
     }
 
+    if (!testReadfile()) {
+        return;
+    }
+
     log.fatal.print("Integration test passed\n");
 }
 
+fn testReadfile() bool {
+    @setRuntimeSafety(false);
+
+    // path_open
+    const fd_addr_in_linear_memory = 0;
+    const path = "test.txt";
+    const path_addr_in_linear_memory = 4;
+    @memcpy(@as([*]u8, @ptrFromInt(linear_memory_offset + path_addr_in_linear_memory)), path);
+    var res = path_open(3, 0, path_addr_in_linear_memory, path.len, 0, 0, 0, 0, fd_addr_in_linear_memory);
+    if (@intFromEnum(res) != 0) {
+        log.fatal.printf("path_open failed: res={d}\n", .{@intFromEnum(res)});
+        return false;
+    }
+    const open_fd = @as(*i32, @ptrFromInt(fd_addr_in_linear_memory + linear_memory_offset));
+
+    // fd_read
+    var buf_iovec_ptr = @as(*IoVec, @ptrFromInt(path_addr_in_linear_memory + 20));
+    buf_iovec_ptr.buf = 32;
+    buf_iovec_ptr.buf_len = 1024;
+    res = fd_read(open_fd.*, @as(i32, @intCast(path_addr_in_linear_memory + 20)), 1, @intCast(path_addr_in_linear_memory + 8 + 20));
+    if (@intFromEnum(res) != 0) {
+        log.fatal.printf("fd_read failed: {d}\n", .{@intFromEnum(res)});
+        return false;
+    }
+
```

Comment:
Could you add verification of the content?

You can use [std.mem.eql](https://ziglang.org/documentation/master/std/#std.mem.eql) to compare two strings.

Reply:
I'm having a problem where I can't find where the data (the contents of test.txt) goes.

```zig

    var iovec_ptr = @as([*]IoVec, @ptrFromInt(@as(usize, @intCast(path_addr_in_linear_memory + 20)) + linear_memory_offset));
    const iovecs = iovec_ptr[0..@as(usize, @intCast(1))];
    const addr = @as(usize, @intCast(iovecs[0].buf)) + linear_memory_offset;
    const received_buf = @as([*]u8, @ptrFromInt(addr))[0..112];
    if (!std.mem.eql(u8, received_buf, path)) {
        log.fatal.printf("content failed, want {s}, get: {s}\n", .{ "fd_read test", received_buf });
        return false;
    }
```
```zig
[LOG DEBUG]: WASI path_open: fd:3, dirflags:0, path_offset:4, path_length:8, oflags:0, rights_bas:0, rights_inferiting:0, fdflags:0, opened_fd_addr:0
[LOG DEBUG]: file path: test.txt
[LOG DEBUG]: WASI fd_read: fd=7 buf_iovec_addr=0x18 vec_len=1 size_addr=0x1c
[LOG FATAL]: content failed, want fd_read test, get: fd_re HTTP/1.1
Host: localhost:1234
User-Agent: curl/7.88.1
Accept: */*

GET / HTTP/1.1
Host: localhost:12
+ exit 1
run ./scripts/integration-test.sh: error: the following command exited with error code 1:
```

Reply:
I've made changes to the original code, but I've run into a new problem.

Run the command: `zig build -Dtest=true -Doptimize=ReleaseFast -Dlog-level=debug run` for testing time. Some times it `succeeds` and some times it `fails`, I don't know why.
```zig
[LOG DEBUG]: WASI path_open: fd:3, dirflags:0, path_offset:4, path_length:8, oflags:0, rights_bas:0, rights_inferiting:0, fdflags:0, opened_fd_addr:0
[LOG DEBUG]: file path: test.txt
[LOG DEBUG]: WASI fd_read: fd=7 buf_iovec_addr=0xc8 vec_len=1 size_addr=0x64
[LOG DEBUG]: WASI fd_close: 7
[LOG FATAL]: Integration test passed
+ QEMU_RETURN_CODE=1
+ RETURN_CODE=0
+ exit 0
+ grep -q 'Integration test passed' build/test/output.txt
+ echo -e '\nIntegration Test PASSED!!'
```

```zig
[LOG DEBUG]: interrupt
[LOG INFO]: client socket test: sock_connect succeeded
[LOG DEBUG]: WASI fd_write: 4 12 2 52
[LOG DEBUG]: interrupt
[LOG DEBUG]: interrupt
[LOG DEBUG]: interrupt
[LOG INFO]: client socket test: fd_write succeeded
[LOG DEBUG]: WASI sock_getlocaladdr: 4 4 40 44
[LOG INFO]: sock_getlocaladdr failed: 112.193.56.4
+ exit 1
run ./scripts/integration-test.sh: error: the following command exited with error code 1:
./scripts/integration-test.sh 
Build Summary: 8/10 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run ./scripts/integration-test.sh failure
error: the following build command failed with exit code 1:
```

Reply:
Oh it's really strange. Does this problem occur only when adding the comparison?

Reply:
> Oh it's really strange. Does this problem occur only when adding the verification of the file content?

It seems that the occurrence doesn't happen with the `testReadfile()` function, the error still occurs after commenting.

![image](https://github.com/mewz-project/mewz/assets/52097565/6f42b8bd-b1d3-4ba3-9d5a-d31178a4269a)

Reply:
So the problem is not triggered by this PR, right?

We might need to open a new issue. But I've not come across this error, so we need to know the condition.

Does it reproduce on Codespaces?

Reply:
Yes, I'm using devcontainer. i get the occasional error.

Reply:
I've confirmed it reproduces on Codespaces on main branch.

So this problem is not related to this PR. I'll create a new issue and merge this PR.